src/hotspot/share/nmt/vmatree.hpp

```diff
@@ -0,0 +1,375 @@
+        rescom.reserve -= B - A;
+      }
+    }
+
+    AddressState prev = {A, stA}; // stA is just filler
+    MEMFLAGS flag_in = LEQ_A.flag();
+    while (to_be_deleted_inbetween_a_b.length() > 0) {
+      const AddressState delete_me = to_be_deleted_inbetween_a_b.top();
+      to_be_deleted_inbetween_a_b.pop();
+      tree.remove(delete_me.address);
+      auto& rescom = diff.flag[NMTUtil::flag_to_index(flag_in)];
+      if (delete_me.state.in == InOut::Reserved) {
+        rescom.reserve -= delete_me.address - prev.address;
+      } else if (delete_me.state.in == InOut::Committed) {
+        rescom.commit -= delete_me.address - prev.address;
+        rescom.reserve -= delete_me.address - prev.address;
+      }
+      prev = delete_me;
+      flag_in = delete_me.flag();
+    }
+    if (prev.address != A &&
+        prev.state.out != InOut::Released &&
+        GEQ_B.state.in != InOut::Released) {
+      // There was some node inside of (A, B) and it is connected to GEQ_B
+      // A - prev - B - GEQ_B
+      // It might be that prev.address == B == GEQ_B.address, this is fine.
+      if (prev.state.out == InOut::Reserved) {
+        auto& rescom = diff.flag[NMTUtil::flag_to_index(prev.flag())];
+        rescom.reserve -= B - prev.address;
+      } else if (prev.state.out == InOut::Committed) {
+        auto& rescom = diff.flag[NMTUtil::flag_to_index(prev.flag())];
+        rescom.commit -= B - prev.address;
+        rescom.reserve -= B - prev.address;
+      }
+    }
+
+    // Finally, we can register the new region [A, B)'s summary data.
+    auto& rescom = diff.flag[NMTUtil::flag_to_index(metadata.flag)];
+    if (state == InOut::Reserved) {
+      rescom.reserve += B - A;
+    } else if (state == InOut::Committed) {
+      rescom.commit += B - A;
+      rescom.reserve += B - A;
+    }
+    return diff;
+  }
+
+  static Metadata identity_merge(Metadata& merge_into, const Metadata& existent) {
+    return merge_into;
+  }
+
+  SummaryDiff reserve_mapping(size_t from, size_t sz, Metadata& metadata) {
+    return register_mapping(from, from + sz, InOut::Reserved, metadata, identity_merge);
+  }
+
+  SummaryDiff commit_mapping(size_t from, size_t sz, Metadata& metadata) {
+    return register_mapping(from, from + sz, InOut::Committed, metadata, [](Metadata& merge_into, const Metadata& existent) {
+      // The committing API takes no flag, so we inherit the flag of the reserved region.
+      merge_into.flag = existent.flag;
+      return merge_into;
+    });
+  }
+
+  SummaryDiff release_mapping(size_t from, size_t sz) {
+    Metadata empty;
+    return register_mapping(from, from + sz, InOut::Released, empty, [](Metadata& merge_into, const Metadata& existent) {
+      // The releasing API takes no flag, so we inherit the flag of the reserved/committed region.
+      // The releasing API also has no call stack, so we inherit the callstack also.
+      merge_into.flag = existent.flag;
+      merge_into.stack_idx = existent.stack_idx;
```

Comment:
Okay, here we don't have to merge at all, no? I don't see any reason to preserve meta info for a released area. 

Therefore, I would just register the area with mtNone and empty-callstack-id.

Reply:
There really was a bug associated with not doing this which caused the invariants of in/out nodes to break, there's a test for it :-). It has to do with the `is_noop() && equals` check. Let's see how this changes as I try out following your idea a bit more.